matchtree.go

```diff
@@ -888,11 +897,93 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
 			all:       regexp.String() == "(?i)(?-s:.)*",
 			matchTree: subMT,
 		}, nil
+
+	case *query.RepoBranches:
+		reposBranchesWant := make(map[uint16]uint64)
+		for _, r := range d.repoMetaData {
+			if branches, ok := s.Set[r.Name]; ok {
+				repoIdx, ok := d.repoMap[r.Name]
+				if !ok {
+					log.Panicf("repoMetaData not in sync with repoMap")
+				}
+				var mask uint64
+				for _, branch := range branches {
+					m, ok := d.branchIDs[repoIdx][branch]
+					if !ok {
+						continue
+					}
+					mask = mask | uint64(m)
+				}
+				if mask != 0 {
+					reposBranchesWant[repoIdx] = mask
+				}
+			}
+		}
+		pred := func(docID uint32) bool {
+			return d.fileBranchMasks[docID]&reposBranchesWant[d.repos[docID]] != 0
+		}
+		return &docMatchTree{
+			docs: d.filterDocs(pred),
+		}, nil
+
+	case *query.RepoSet:
+		var reposWant map[uint16]struct{}
+		for _, r := range d.repoMetaData {
+			if _, ok := s.Set[r.Name]; ok {
+				repoWant, ok := d.repoMap[r.Name]
+				if !ok {
+					log.Panicf("repoMetaData not in sync with repoMap")
+				}
+				reposWant[repoWant] = struct{}{}
+			}
+		}
+		// Optimization: If all repos within a shard are to be searched we can just
+		// return a bruteForceMatchTree.
+		if len(reposWant) == len(d.repoMetaData) {
+			return &bruteForceMatchTree{}, nil
+		}
```

Comment:
I think this optimization is more appropriate in `indexData.simplify` since you then also get constant folding afterwards.

Reply:
I moved this to simplify and created a separate function 'simplifyMultiRepo' to reuse code.